l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,345 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.20;
+
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+import {EfficientCall} from "@matterlabs/zksync-contracts/l2/system-contracts/libraries/EfficientCall.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @title ConsensusRegistry
+/// @dev Manages consensus nodes and committees for the L2 consensus protocol,
+/// owned by a governance protocol. Nodes act as both validators and attesters,
+/// each playing a distinct role in the consensus process. This contract facilitates
+/// the rotation of validator and attester committees, which represent a subset of nodes
+/// expected to actively participate in the consensus process during a specific time window.
+contract ConsensusRegistry is Ownable2Step {
+    // An array to keep track of node owners.
+    address[] public nodeOwners;
+    // A map of node owners => nodes.
+    mapping(address => Node) public nodes;
+    // The current validator committee list.
+    CommitteeValidator[] public validatorCommittee;
+    // The current attester committee list.
+    CommitteeAttester[] public attesterCommittee;
+
+    /// @dev Represents a consensus node.
+    struct Node {
+        // A flag stating if the node is active.
+        // Inactive nodes are not considered when selecting committees.
+        bool active;
+        // Validator's voting weight.
+        uint256 validatorWeight;
+        // Validator's BLS12-381 public key.
+        bytes validatorPubKey;
+        // Validator's Proof-of-possession (a signature over the public key).
+        bytes validatorPoP;
+        // Attester's Voting weight.
+        uint256 attesterWeight;
+        // Attester's ECDSA public key.
+        bytes attesterPubKey;
+    }
+
+    /// @dev Represents a validator committee member.
+    struct CommitteeValidator {
+        address nodeOwner;
+        uint256 weight;
+        bytes pubKey;
+        bytes pop;
+    }
+
+    /// @dev Represents an attester committee member.
+    struct CommitteeAttester {
+        uint256 weight;
+        address nodeOwner;
+        bytes pubKey;
+    }
```

Comment:
@vladbochok The pub keys and the weights are duplicated here because they can be changed on the node struct at any time but shouldn't change on the committee until we call `setAttesterCommittee`. So the `attesterCommittee` and `validatorCommittee` are meant to be snapshots of the attesters/validators at a given time.
We can maybe have some notion of old and new node. Changes to the nodes would be to the "new" node struct, while the committee would point to the "old" node struct. Calling `setAttesterCommittee` would switch the old and new nodes.
Besides that, we can have some uint32 node id instead of using an address. That saves some more space. And then the nodeOwner address could be mutable, which is an advantage.
With both solutions, we could have `attesterCommittee` and `validatorCommittee` be just an array of uint32.

Reply:
I guess we can further simply and dispense with the notion of committee lists decoupled from the node list. We can add the following fields to `Node` to capture updates:

* `pendingValidatorActive`
* `pendingValidatorPubKey`
* `pendingValidatorPoP`
* `pendingValidatorWeight`
* `pendingAttesterActive`
* `pendingAttesterPubKey`
* `pendingAttesterWeight`

Then we’ll have `commitValidators` / `commitAttesters` to iterate through nodes and commit to non-empty pending values. 

I'm not sure how much it will improve the storage costs. Also, it might justify using dynamic-size types to indicate "null" pending values. But it seems to me like the most minimal way to achieve pending/commit scheme on storage.

Reply:
It still iterating over array, I would do rather this way:
- contract will have additional storage value `epoch` or `snapshotNumber`
- When Owner wants to commit `commitValidators` and `commitAttesters` it will increase the 
`epoch` value
- each node will have an additional field `lastUpdatedAtEpoch` 

Now offchain observer can easily restore the list of validators and attesters. WDYT?

Reply:
That still requires nodes to have history. Nodes currently sync from a snapshot (so they have no history at all) and then in order to be able to sync the rest of the chain they need to know the attester/validator list for the next batch. That's why we need this "duplication" of data.

Reply:
> Note that different selection mechanisms (like PoS) will only pick a subset of all nodes, possibly at random.

I assume it can be done by adding an indicator field similar to `active`, which offchain consensus code will use. It can be generated on storage while doing the commit. 

> Also, we want at a given point to cache lists for several batches (but we don't need to worry about it for now).

If there’s no commit, then it’s a defacto cache. Additional delays can be applied offchain.

Reply:
To clarify, I don't think what I described here https://github.com/matter-labs/era-contracts/pull/555#discussion_r1705578094 is a great improvement over what we have now. 

Given that we don't want to read from history, fields will need to be cloned this way or another.

Reply:
> I assume it can be done by adding an indicator field similar to active, which offchain consensus code will use.
If we're just going to add a in-committee flag to each node, then it's basically the same as having a list attesters. Except that it occupies less space (1 byte for a boolean vs 4 for a u32).

Reply:
I agree with keeping separate lists. Let's just try to reduce as much as possible the gas used to update the lists, and instead move the necessary computation to when we add/modify nodes.

Reply:
@brunoffranca @moshababo Why do you need a separate list?...

What I tried to describe is the lazy update technique.  
- https://github.com/matter-labs/era-contracts/pull/555#discussion_r1705613987

> That still requires nodes to have history.

You need to remember only two node attributes the active one and the cached one (from the previous epoch). This is better than updating the list by recopy it because you only 
1) pay gas when the node attributes changed
2) You pay for each node update separately rather than once for all of them (no DoS function)

Reply:
If you're a new node syncing up you would need potentially to have all history to get the active node attributes. Because on the contract we only have the current attributes.
This is a non-starter, ENs sync using snapshots by default now, so they start with no history at all. Just a state snapshot.
Given this, there is no way we can avoid having both attributes on storage.

Reply:
> If you're a new node syncing up you would need potentially to have all history to get the active node attributes.
Why?...

> Because on the contract we only have the current attributes.

But I just said about saving current and previous attributes, that's it. It is enough to sync, isn't it?

Reply:
> Also, the advantages you mention can be achieved by mine or Moshe's suggestions also. The old and new attributes are kept in the node, and the committee lists are just pointers to the nodes.

If you mean this suggestion:
- https://github.com/matter-labs/era-contracts/pull/555#discussion_r1705578094

Then you still have O(n) writes per list updates. Not very good. In my suggestion we can remove that

Reply:
> I think we are misunderstanding each other. First, in your proposal, does the Node struct store both active and pending attributes?

Yes, it stores both.

> Second, in your proposal, how do you know which attesters are part of the committee and which ones are not?

That being said, we also store the `epoch` number as an additional storage variable. Also, each node will have one more field `lastUpdatedAtEpoch`. When the owner changes attributes on the pending node, the contract firstly check whether the current `epoch` is bigger than `lastUpdatedAtEpoch` on node, then makes the pending node an active node, and only then makes changes to the pending one.

Offchain actor can read the current `epoch` and all nodes. If the `lastUpdatedAtEpoch` is equal to the current node just use active one, and if  `lastUpdatedAtEpoch` is less then `epoch` use pending node.

Reply:
> Yes, it stores both.

Ok, that's fine then. I was imagining a different algorithm. Then I have no objections. Later we will need to cache several epochs which will complicate this scheme, but we can figure it out then.
@moshababo if you have no concerns, we can go with this scheme.

Reply:
@vladbochok just to clarify, will there be a need to double the storage capacity, to maintain 2 distinct sets of nodes (pending and active), and `lastUpdatedAtEpoch` will need to maintained once for both (perhaps in a separate structure)?

Reply:
> will there be a need to double the storage capacity, to maintain 2 distinct sets of nodes

Yes, the node struct can basically store two fields `pendingNodeAttr` and `activeNodeAttr`.

> lastUpdatedAtEpoch will need to be maintained once for both for each node (perhaps in a separate structure)?

Can be in the same structure in fact:

```
struct NodeAttr {
        // A flag stating if the node is active.
        // Inactive nodes are not considered when selecting committees.
        bool active;
        // Validator's voting weight.
        uint32 validatorWeight;
        // Validator's BLS12-381 public key.
        BLS12_381PublicKey validatorPubKey;
        // Validator's Proof-of-possession (a signature over the public key).
        BLS12_381Signature validatorPoP;
        // Attester's Voting weight.
        uint32 attesterWeight;
        // Attester's Secp256k1 public key.
        Secp256k1PublicKey attesterPubKey;
}

struct Node {
        NodeAttr pendingNode;
        NodeAttr activeNode;
        uint256 lastUpdatedAtEpoch;
}
```

Reply:
I pushed some code [here](https://github.com/matter-labs/era-contracts/pull/555/commits/c13bd81974c0fbcb284328be6e10c739b77999e5), to get feedback before finalizing.

Main remarks:
* Renamed `pending`/`active` as `latest`/`snapshot`. The former did not make sense to me since both modes can be used to read from.
* Renamed `epoch` as `commit`. The former is too overloaded in the context of consensus and isn't correlated with its common meaning.
* When adding nodes, `snapshot` is assigned zero value.
* Attester and Validator attributes were decoupled, since they need to be committed to separately.
* `active` attribute was placed in both, separately, for the same reason.
* Removal of nodes became more complicated. I applied 3 phases: first is the scheduling, where the removal is pending commits for both attesters and validators. Once one of them is committed to, the node is considered as removed for that role. Once the second role is committed to, the removal is finalized with the actual removal from storage.